File: vm/tests/all/interop/lambdaExpr/testMain.cpp
//lambdaExpr.cpp
#include<polyglot.h>
#include<iostream>
using namespace std;

int doubleAndIncrement(int x) {
	return x*2+1;
}

int main() {

PolyglotFile* llFile = Polyglot.evalFile("llvm", "lambdaExprClass.so");
PolyglotFile* jsFile = Polyglot.evalFile("js", "lambdaExprClass.js");
PolyglotFile* pyFile = Polyglot.evalFile("python", "lambdaExprClass.py");

int nErrors=0;

LLVM_MyClass* c = new llFile->invoke("MyClass");
int val = c->foo(8);
if(val!=8) {
	nErrors++;
	cout << "C++/LLVM: calling f(8, x->x) leads to " << val << endl;
}
c->foo = doubleAndIncrement;
val = c->foo(8);
if(val!=17) {
	nErrors++;
	cout << "C++/LLVM: calling f(8, x->x) leads to " << val << endl;
}

JS_MyClass* cJ = new jsFile->invoke("MyClass");
val = cJ->foo(8);
if(val!=8) {
	nErrors++;
	cout << "JS: calling f(8, x->x) leads to " << val << endl;
}
c->foo = doubleAndIncrement;
val = cJ->foo(8);
if(val!=17) {
	nErrors++;
	cout << "JS: calling f(8, x->x) leads to " << val << endl;
}

Python_MyClass* cP = new pyFile->invoke("MyClass");
val = cP->foo(8);
if(val!=8) {
	nErrors++;
	cout << "Python: calling f(8, x->x) leads to " << val << endl;
}
c->foo = doubleAndIncrement;
val = cP->foo(8);
if(val!=17) {
	nErrors++;
	cout << "Python: calling f(8, x->x) leads to " << val << endl;
}

if(nErrors==0) {
	cout << "OK" << endl;
} else {
	cout << nErrors << " errors." << endl;
}

delete cP;
delete cJ;
delete c;

delete llFile;
delete jsFile;
delete pyFile;
}

File: vm/tests/all/interop/lambdaExpr/testMain.js
//lambdaExpr.js

function check(nErrors, info, expected, actual) {
	if(expected!=actual) {
		console.log(info, actual);
		nErrors++;
	}
	return nErrors;
}

var llFile = Polyglot.evalFile("llvm", "lambdaExprClass.so");
var jsFile = Polyglot.evalFile("js", "lambdaExprClass.js");
var pyFile = Polyglot.evalFile("python", "lambdaExprClass.py");

nErrors = 0;

c = new llFile.MyClass();
nErrors = check(nErrors, "calling f(8, x->x) leads to ", 8, c.foo(8));
c.foo = (x => x*2+1);
nErrors = check(nErrors, "calling f(8, x->2*x+1) leads to ", 17, c.foo(8));

cJ = new jsFile.MyClass();
nErrors = check(nErrors, "calling f(8, x->x) leads to ", 8, cP.foo(8));
cJ.foo = (x => x*2+1);
nErrors = check(nErrors, "calling f(8, x->2*x+1) leads to ", 17, cJ.foo(8));

cP = new pyFile.MyClass();
nErrors = check(nErrors, "calling f(8, x->x) leads to ", 8,  cP.foo(8));
cP.foo = (x => x*2+1);
nErrors = check(nErrors, "calling f(8, x->2*x+1) leads to ",17,  cP.foo(8));

if(nErrors==0) {
	console.log("OK");
} else {
	console.log(nErrors, " errors.");
}
